parser/src/parser.rs

```diff
@@ -1103,4 +1126,65 @@ class Abcd:
         .unwrap();
         insta::assert_debug_snapshot!(parse_ast);
     }
+
+    #[test]
+    fn test_jupyter_magic() {
+        let parse_ast = parse(
+            r#"
+# Normal Python code
+(
+    a
+    %
+    b
+)
+
+# Dynamic object info
+??a.foo
+?a.foo
+?a.foo?
+??a.foo()??
+
+# Line magic
+%timeit a = b
+%timeit foo(b) % 3
+%alias showPath pwd && ls -a
+%timeit a =\
+  foo(b); b = 2
+%matplotlib --inline
+%matplotlib \
+    --inline
+
+# System shell access
+!pwd && ls -a | sed 's/^/\    /'
+!pwd \
+  && ls -a | sed 's/^/\\    /'
+!cd /Users/foo/Library/Application\ Support/
+
+# Let's add some Python code to make sure that earlier escapes were handled
+# correctly and that we didn't consume any of the following code as a result
+# of the escapes.
+def foo():
+    return (
+        a
+        !=
+        b
+    )
+
+# Transforms into `foo(..)`
+/foo 1 2
+;foo 1 2
+,foo 1 2
+
+# Indented magic
+for a in range(5):
+    %ls
+    pass
```

Comment:
Can we add some cases that look like magics but are none:

```python
a % 5

(
	a
	% 
	5
)

%

a \
	% 5


```

Reply:
> ```python
> a % 5
> 
> (
> 	a
> 	% 
> 	5
> )
> 
> %  # [*]
> 
> a \
> 	% 5
> ```

They should all work assuming that for the marked (`[*]`) line, the `%` sign is a standalone (not part of another expression). It isn't as Python throws a Syntax error on that line and the tokenizer gives the `MagicCommand` token only for that line.

Reply:
Ah, this all works because we only call into `lex_indentation` when it is *not* parenthesized.

Reply:
Yes that's correct.